sgkit_plink/pysnptools.py

```diff
@@ -0,0 +1,252 @@
+"""PLINK 1.9 reader implementation"""
+from pathlib import Path
+from typing import Union
+
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+from dask.dataframe import DataFrame
+from pysnptools.snpreader import Bed
+from xarray import Dataset
+
+from sgkit import create_genotype_call_dataset
+from sgkit.api import DIM_SAMPLE
+
+PathType = Union[str, Path]
+
+FAM_FIELDS = [
+    ("family_id", str, "U"),
+    ("member_id", str, "U"),
+    ("paternal_id", str, "U"),
+    ("maternal_id", str, "U"),
+    ("sex", str, "int8"),
+    ("phenotype", str, "int8"),
+]
+FAM_DF_DTYPE = dict([(f[0], f[1]) for f in FAM_FIELDS])
+FAM_ARRAY_DTYPE = dict([(f[0], f[2]) for f in FAM_FIELDS])
+
+BIM_FIELDS = [
+    ("contig", str, "U"),
+    ("variant_id", str, "U"),
+    ("cm_pos", "float32", "float32"),
+    ("pos", "int32", "int32"),
+    ("a1", str, "U"),
+    ("a2", str, "U"),
+]
+BIM_DF_DTYPE = dict([(f[0], f[1]) for f in BIM_FIELDS])
+BIM_ARRAY_DTYPE = dict([(f[0], f[2]) for f in BIM_FIELDS])
+
+
+class BedReader(object):
+    def __init__(self, path, shape, dtype=np.int8, count_A1=True):
+        # n variants (sid = SNP id), n samples (iid = Individual id)
+        n_sid, n_iid = shape
+        # Initialize Bed with empty arrays for axis data, otherwise it will
+        # load the bim/map/fam files entirely into memory (it does not do out-of-core for those)
+        self.bed = Bed(
+            str(path),
+            count_A1=count_A1,
+            # Array (n_sample, 2) w/ FID and IID
+            iid=np.empty((n_iid, 2), dtype="str"),
+            # SNP id array (n_variants)
+            sid=np.empty((n_sid,), dtype="str"),
+            # Contig and positions array (n_variants, 3)
+            pos=np.empty((n_sid, 3), dtype="int"),
+        )
+        self.shape = (n_sid, n_iid, 2)
+        self.dtype = dtype
+        self.ndim = 3
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, tuple):
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be tuple (received {type(idx)})"
+            )
+        if len(idx) != self.ndim:
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be two-item tuple (received {len(idx)} slices)"
+            )
+        # Slice using reversal of first two slices --
+        # pysnptools uses sample x variant orientation
+        arr = self.bed[idx[1::-1]].read(dtype=np.float32, view_ok=False).val.T
+        # Convert missing calls as nan to -1
+        arr = np.nan_to_num(arr, nan=-1.0)
+        arr = arr.astype(self.dtype)
+        # Add a ploidy dimension, so allele counts of 0, 1, 2 correspond to 00, 01, 11
+        arr = np.stack(
+            [
+                np.where(arr < 0, -1, np.where(arr == 0, 0, 1)),
+                np.where(arr < 0, -1, np.where(arr == 2, 1, 0)),
+            ],
+            axis=-1,
+        )
```

Comment:
This is tricky code and warrants a unit test. For an allele count of 1 this returns [1, 0], not [0, 1], which seems to be the wrong way round...

Reply:
Why does the order matter if it's always unphased data with PLINK?  I thought it made more sense for an unphased heterozygous call to have a 1 in the first dimension so that it's easier to identify homozygous reference calls, but I'm not aware of that having any other significance.  I also that thought allele counting and missingness validations were sufficient tests given that those are all the things relevant for that bit of code, but I could validate individual values in 'call/genotype' as well.  That's what you mean right?

Reply:
It's true that for unphased data 01 and 10 are the same. The comment says 01, so it's slightly inconsistent with that.

> I also that thought allele counting and missingness validations were sufficient tests given that those are all the things relevant for that bit of code, but I could validate individual values in 'call/genotype' as well. That's what you mean right?

Yes that's what I meant - check a few values.

Reply:
Ah ok, I thought you meant something else by wrong way round.  I fixed the comment and added a [test](https://github.com/pystatgen/sgkit-plink/pull/1/commits/ee717081c01e1a1f635c027909f559f78d076230#diff-33e9afa89862f603b25f9c5abf5ef334R37).  Will resolve if that makes sense to you.